### PR TITLE
Feature_1716:set registry and version variable in deploy-karmada.sh

### PR DESCRIPTION
**Pull request by wuyingjun-lucky**
Signed-off-by: wuyingjun <wuyingjun_yewu@cmss.chinamobile.com>

**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:
Fixes # https://github.com/karmada-io/karmada/issues/1716
**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

### CODE REVIEW

**Comment by lonelyCZ** [comment]
Thanks @wuyingjun-lucky , I will look it ASAP.

File: hack/deploy-scheduler-estimator.sh [hunk 1]
```diff
@@ -4,6 +4,12 @@ set -o errexit
 set -o nounset
 
 REPO_ROOT=$(dirname "${BASH_SOURCE[0]}")/..
+
+#define karmada control plane image registry and version
+KARMADA_REGISTRY=${KARMADA_REGISTRY:-"swr.ap-southeast-1.myhuaweicloud.com/karmada"}
+KARMADA_VERSION=${KARMADA_VERSION:-"latest"}
+KUBE_REGISTRY=${KUBE_REGISTRY:-"k8s.gcr.io"}
```
Comment (lonelyCZ) [inline hack/deploy-scheduler-estimator.sh:11]: We seemingly don't use this variate `KUBE_REGISTRY` in this script.

**Comment by lonelyCZ** [comment]
Should we put this parameters in `function usage()`

https://github.com/karmada-io/karmada/blob/dd75a65f1014460652018e60e0cbf4c74a694da4/hack/deploy-karmada.sh#L22-L32

File: hack/deploy-scheduler-estimator.sh (hunk 1 above)
Comment (wuyingjun-lucky) [reply hack/deploy-scheduler-estimator.sh:11]: GOT IT

File: hack/deploy-karmada.sh [hunk 2]
```diff
@@ -12,6 +12,11 @@ mkdir -p "${CERT_DIR}" &>/dev/null ||  mkdir -p "${CERT_DIR}"
 rm -f "${CERT_DIR}/*" &>/dev/null ||  rm -f "${CERT_DIR}/*"
 KARMADA_APISERVER_SECURE_PORT=${KARMADA_APISERVER_SECURE_PORT:-5443}
 
+# Define karmada control plane image registry and version
+KARMADA_REGISTRY=${KARMADA_REGISTRY:-"swr.ap-southeast-1.myhuaweicloud.com/karmada"}
+KARMADA_VERSION=${KARMADA_VERSION:-"latest"}
+KUBE_REGISTRY=${KUBE_REGISTRY:-"k8s.gcr.io"}
```
Comment (lonelyCZ) [inline hack/deploy-karmada.sh:18]: I like the variate `KUBE_REGISTRY` very much that reduces my steps of deployment~

cc @RainbowMango

**Comment by lonelyCZ** [comment]
> 1: I found other shell scripts or make file have similar env named VERSION and REGISTRY;
should I use VERSION and REGISTRY and KUBE_REGISTRY in my pull request or change them to KARMADA_VERSION and KARMADA_REGISTRY and KUBE_REGISTRY for the code and env consistency

I think it can be unified to `VERSION`, `REGISTRY` and `KUBE_REGISTRY`. `deploy-karmada.sh` is  more low-level and they don't conflict.

> 2: local-up.sh use the deploy-karmada.sh to deploy clusters; But when I set env CHINA_MAINLAND the scripts will replace the registry value like this

Right, perhaps we should fix the function `CHINA_MAINLAND`, we should not update the files in `REPO_ROOT`.

**Comment by lonelyCZ** [comment]
I find that if we modify REGISTRY here, can we not deploy it successfully? It appears that the yaml file below has not changed

https://github.com/karmada-io/karmada/blob/dd75a65f1014460652018e60e0cbf4c74a694da4/hack/local-up-karmada.sh#L122-L130

**Comment by wuyingjun-lucky** [comment]
> I find that if we modify REGISTRY here, can we not deploy it successfully? It appears that the yaml file below has not changed
> 
> https://github.com/karmada-io/karmada/blob/dd75a65f1014460652018e60e0cbf4c74a694da4/hack/local-up-karmada.sh#L122-L130

Yes , I found it today when test local-up.sh   and I will fix it  like this 
image
do u think it ok ?

**Comment by lonelyCZ** [comment]
I feel good about making the script more flexible. Perhaps we can look forward to @RainbowMango comments on such changes.

**Comment by wuyingjun-lucky** [comment]
> I feel good about making the script more flexible. Perhaps we can look forward to @RainbowMango comments on such changes.

Ok, before that I will modify the code with your suggestion ; And please help me again

**Comment by wuyingjun-lucky** [comment]
@RainbowMango Could you help me review the pull request ？

**Comment by RainbowMango** [comment]
Sure.
/assign

**Comment by RainbowMango** [comment]
Generally looks good to me. 
But I guess I guess the `deploy-karmada.sh` is mainly used by developers, if you want to install Karmada it is recommended to refer to [Installation Guide](https://github.com/karmada-io/karmada/blob/master/docs/installation/installation.md).

**Comment by RainbowMango** [comment]
@lfbear Could you please take a look?

**Comment by RainbowMango** [comment]
I see you did excellent work here. My concern is if this will make the scripts too complicated?
Can you guys give me more hints about the benefits we got?

**Comment by wuyingjun-lucky** [comment]
> I see you did excellent work here. My concern is if this will make the scripts too complicated? Can you guys give me more hints about the benefits we got?
I do not think this the scripts is too complicated，you can use the scripts to deploy as before，we dont change the process, just set it variable。
our team use it  frequently;  we had to pull the code and then modify the scripts to deploy clusters time and time. But we dont want to do this. we want our code is not dirty.
I guess people who use the swr.ap-southeast-1.myhuaweicloud.com may think it is not necessary, But people who don not use swr.ap-southeast-1.myhuaweicloud.com and  k8s.gcr.io will welcome that

**Comment by RainbowMango** [comment]
> our team use it frequently; we had to pull the code and then modify the scripts to deploy clusters again and again.

May I know why you are so dependent on `deploy-karmada.sh`?  (not an objection, just curious).
You can use `karmadactl init` to install Karmada, sure you can customize the images registry and version.
Or use `local-up-karmada.sh` to install Karmada from the source, in this case you don't have to replace registry.

**Comment by wuyingjun-lucky** [comment]
@RainbowMango Should I close the change or modify the change ？

**Comment by wuyingjun-lucky** [comment]
I have tested the kubectl-karmada. 
It can not used in our environment，because it had to download some crd files from internet. 
It can not be used by a offline scheme. @RainbowMango

**Comment by chaunceyjiang** [comment]
Maybe this is a more elegant [solution](https://github.com/karmada-io/karmada/issues/1753)

**Pull request closed without merging**